### ISSUE TITLE
Hello, may I ask how to render depth images based on the environment of this project?

### ISSUE THREAD

**Issue by 3513695749** [windrise/3DGR-CAR#5]

**Comment by windrise** [comment]
In this work, we utilize normalized depth maps for auxiliary supervision of 3D point prediction from a single view, which are computed using the following function:
def voxel_to_depth_map(voxel_data):
    """
    Convert 3D voxel data into a depth map along the y-axis direction.

    Parameters:
    - voxel_data: 3D voxel data with shape (1, x, y, z), assumed to be mostly zeros 
                  with non-zero values representing structures.

    Returns:
    - depth_map: A depth map with shape (x,y) representing the structural depth 
                 when viewed from the negative to positive y-axis direction.
    """
    # Assume voxel data has shape (1, D, H, W), remove batch dimension
    voxel_data = voxel_data.squeeze(0)

    # Initialize depth map with zeros, shape (H, W)
    depth_map = torch.ones(voxel_data.shape[0], voxel_data.shape[2])

    # Iterate through each position on the x-z plane
    for x in range(voxel_data.shape[0]):
        for z in range(voxel_data.shape[2]):
            # Search for the first non-zero voxel along y-axis
            for y in range(voxel_data.shape[1]):
                if voxel_data[x, y, z] > 0:
                    # Calculate distance from non-zero voxel to x0z plane (y value)
                    depth_map[x, z] = y/voxel_data.shape[1]
                    break  # Exit loop after finding first non-zero value

    return depth_map

**Comment by 3513695749** [comment]
> In this work, we utilize normalized depth maps for auxiliary supervision of 3D point prediction from a single view, which are computed using the following function: def voxel_to_depth_map(voxel_data): """ Convert 3D voxel data into a depth map along the y-axis direction.
> 
> ```
> Parameters:
> - voxel_data: 3D voxel data with shape (1, x, y, z), assumed to be mostly zeros 
>               with non-zero values representing structures.
> 
> Returns:
> - depth_map: A depth map with shape (x,y) representing the structural depth 
>              when viewed from the negative to positive y-axis direction.
> """
> # Assume voxel data has shape (1, D, H, W), remove batch dimension
> voxel_data = voxel_data.squeeze(0)
> 
> # Initialize depth map with zeros, shape (H, W)
> depth_map = torch.ones(voxel_data.shape[0], voxel_data.shape[2])
> 
> # Iterate through each position on the x-z plane
> for x in range(voxel_data.shape[0]):
>     for z in range(voxel_data.shape[2]):
>         # Search for the first non-zero voxel along y-axis
>         for y in range(voxel_data.shape[1]):
>             if voxel_data[x, y, z] > 0:
>                 # Calculate distance from non-zero voxel to x0z plane (y value)
>                 depth_map[x, z] = y/voxel_data.shape[1]
>                 break  # Exit loop after finding first non-zero value
> 
> return depth_map
> ```
Thank you for your quick reply. The code you provided works. However, I am not sure how to generate the depth map corresponding to "proj_data" in the "forward_project" function in the "ConeBeam3DProjector" class. Do you know the solution?

**Comment by windrise** [comment]
Sorry I wasn't at my desk when I saw the message. Because for us, we obtain the rotated depth map and corresponding projection image by first voxelizing the point cloud, then rotating the point cloud, and finally converting the rotated image back to voxels. You can refer to the code hint below:
`
Reference 

    save_dir = r"/path/to/dataset_save"
    image_dir = r"/path/to/dataset"
    for file in CASData_list[args.n1:args.n2]:
        print("filename: ", file)
        file_path = os.path.join(image_dir, str(file), 'img.nii.gz')
        label_path = os.path.join(image_dir, str(file), 'label.nii.gz')
        image = nib.load(file_path)
        labels = nib.load(label_path)
        image_mha = image.get_fdata() * (labels.get_fdata() == 1)
        image_mha[image_mha < 0] = 0

        image_mha = crop_to_standard_cas(image_mha, scale=480)  # ImageCAS
        labels_cor = crop_to_standard_cas(labels.get_fdata(), scale=480)  # ImageCAS
        image_mha = resample_valid_voxels_to_standard(image_mha, 128)
        labels_cor = resample_valid_voxels_to_standard(labels_cor , 128)

        mask = labels_cor == 0
        image_mha = torch.tensor(image_mha, dtype=torch.float32)[None, ...]
        mask = torch.tensor(mask, dtype=torch.float32)[None, ...]
        image_mha = F.interpolate(image_mha, size=(128, 128), mode='bilinear', align_corners=False)
        resampled_mask = F.interpolate(mask, size=(128, 128), mode='bilinear', align_corners=False)
        image_mha[resampled_mask.bool()] = 0

        image_mha = image_mha / torch.max(image_mha)  # [1, C, H, W]

        voxel_center = torch.tensor([64,64,64], dtype=torch.float32)

        #label
        points = torch.argwhere(1 - resampled_mask.squeeze(0))
        points_voxels = 1 - resampled_mask.squeeze(0)
        #centerline   
        centerline_voxels = torch.tensor(skeletonize_3d(points_voxels)).long()
        centerline_points = np.argwhere(centerline_voxels).T
        number = 360
        bg_mask = torch.zeros(number , points.shape[0],3)
        depthes = torch.zeros(number , 128,128)
        cl_mask = torch.zeros(number , centerline_points.shape[0],3)
        projs_all = torch.zeros(number , 128, 128)
        points_centered = points - voxel_center
        centerline_points_centered = centerline_points - voxel_center
        #rotate resampled_mask
        points_image = volume2pointsplusplus(image_mha)
        points_image_center = points_image[:,:3] - voxel_center
        points_value = points_image[:, 3].unsqueeze(1)
        volume = torch.zeros(number , 1, 128, 128, 128, dtype=points_image.dtype, device=points_image.device)
        # parallel computing
        from multiprocessing import Pool
        def process(i):
            # Perform rotation operations on the point cloud
            bg_mask = (rotate_coordinate_system(points_centered, i) + voxel_center) / 128
            cl_mask = (rotate_coordinate_system(centerline_points_centered, i) + voxel_center) / 128
            # 将Convert the rotated point cloud back to voxels
            points_image_rotated = rotate_coordinate_system(points_image_center, i) + voxel_center
            points_image_rotated = torch.cat((points_image_rotated / 128, points_value), dim=1)
            rotated_image = pointsplusplus2volume(points_image_rotated, 128)
            projs = ct_projector.forward_project(rotated_image).cpu()
            projs_all = torch.flip(projs.squeeze(0)[0, :, :], [1]).T
            dep = voxel_to_depth_map(rotated_image)
            depthes = torch.flip(dep, [1]).T
            volume = rotated_image
            return projs_all, bg_mask, cl_mask, depthes, volume
        with Pool() as p:
            results = p.map(process, range(number))

        projs_all, bg_mask, cl_mask, depthes, volume = zip(*results)

        #supplement 
        def rotation_matrix_z(theta):
            theta = torch.tensor(theta, dtype=torch.float32) * torch.pi / 180.0
            cos_theta = torch.cos(theta)
            sin_theta = torch.sin(theta)
            return torch.tensor([
                [cos_theta, -sin_theta, 0],
                [sin_theta, cos_theta, 0],
                [0, 0, 1]
            ], dtype=torch.float32)
        
        def rotate_coordinate_system(points, theta):
            R = rotation_matrix_z(theta)
            return torch.mm(points, R.t())

        def pointsplusplus2volume(pointsplusplus, volume_size=128):
            pointsplusplus[:, :3] = (pointsplusplus[:, :3] * (volume_size - 1)).long()
            volume = torch.zeros(1, volume_size, volume_size, volume_size, dtype=pointsplusplus.dtype, device=pointsplusplus.device)
            coordinates = pointsplusplus[:, :3].long()
            values = pointsplusplus[:, 3]
            volume[0, coordinates[:, 0], coordinates[:, 1], coordinates[:, 2]] = values
            return volume
        def volume2pointsplusplus(volume):
            # volume [1, depth, height, width]
            volume = volume.squeeze(0)
            points = torch.argwhere(volume)
            values = volume[points[:, 0], points[:, 1], points[:, 2]].unsqueeze(-1)
            pointsplusplus = torch.cat((points, values), dim=1)
            return pointsplusplus

        def resample_valid_voxels_to_standard(scan, scale):
            z, y, x = scan.shape
            z_scale, y_scale, x_scale = scale / z, scale / y, scale / x
            background_mask = scan == 0
            resampled_scan = zoom(scan, (z_scale, y_scale, x_scale), order=1, mode='nearest')
            resampled_background_mask = zoom(background_mask.astype(int), (z_scale, y_scale, x_scale), order=0, mode='nearest').astype(bool)
            resampled_scan[resampled_background_mask] = 0
            return resampled_scan

        def crop_to_standard_cas(source_scan, scale):
        
            scan = recenter_voxels(source_scan)
            x, y, z = scan.shape
            if z >= scale:
                start_z = (z - scale) // 2
                ret_scan = scan[:, :, start_z:start_z + scale]
            else:
                temp1 = np.zeros((x, y, (scale - z) // 2))
                temp2 = np.zeros((x, y, (scale - z) - (scale - z) // 2))
                ret_scan = np.concatenate((temp1, scan, temp2), axis=2)
        
            x, y, z = ret_scan.shape
        
            if y >= scale:
                start_y = (y - scale) // 2
                ret_scan = ret_scan[:, start_y:start_y + scale, :]
            else:
                temp1 = np.zeros((x, (scale - y) // 2, z))
                temp2 = np.zeros((x, (scale - y) - (scale - y) // 2, z))
                ret_scan = np.concatenate((temp1, ret_scan, temp2), axis=1)
        
            x, y, z = ret_scan.shape
        
            if x >= scale:
                start_x = (x - scale) // 2
                ret_scan = ret_scan[start_x:start_x + scale, :, :]
            else:
                temp1 = np.zeros(((scale - x) // 2, y, z))
                temp2 = np.zeros(((scale - x) - (scale - x) // 2, y, z))
                ret_scan = np.concatenate((temp1, ret_scan, temp2), axis=0)
        
            return ret_scan
`
You need to carefully verify the code and check whether the intermediate outputs match the expected results.

**Comment by 3513695749** [comment]
> Sorry I wasn't at my desk when I saw the message. Because for us, we obtain the rotated depth map and corresponding projection image by first voxelizing the point cloud, then rotating the point cloud, and finally converting the rotated image back to voxels. You can refer to the code hint below: ` Reference
> 
> ```
> save_dir = r"/path/to/dataset_save"
> image_dir = r"/path/to/dataset"
> for file in CASData_list[args.n1:args.n2]:
>     print("filename: ", file)
>     file_path = os.path.join(image_dir, str(file), 'img.nii.gz')
>     label_path = os.path.join(image_dir, str(file), 'label.nii.gz')
>     image = nib.load(file_path)
>     labels = nib.load(label_path)
>     image_mha = image.get_fdata() * (labels.get_fdata() == 1)
>     image_mha[image_mha < 0] = 0
> 
>     image_mha = crop_to_standard_cas(image_mha, scale=480)  # ImageCAS
>     labels_cor = crop_to_standard_cas(labels.get_fdata(), scale=480)  # ImageCAS
>     image_mha = resample_valid_voxels_to_standard(image_mha, 128)
>     labels_cor = resample_valid_voxels_to_standard(labels_cor , 128)
> 
>     mask = labels_cor == 0
>     image_mha = torch.tensor(image_mha, dtype=torch.float32)[None, ...]
>     mask = torch.tensor(mask, dtype=torch.float32)[None, ...]
>     image_mha = F.interpolate(image_mha, size=(128, 128), mode='bilinear', align_corners=False)
>     resampled_mask = F.interpolate(mask, size=(128, 128), mode='bilinear', align_corners=False)
>     image_mha[resampled_mask.bool()] = 0
> 
>     image_mha = image_mha / torch.max(image_mha)  # [1, C, H, W]
> 
>     voxel_center = torch.tensor([64,64,64], dtype=torch.float32)
> 
>     #label
>     points = torch.argwhere(1 - resampled_mask.squeeze(0))
>     points_voxels = 1 - resampled_mask.squeeze(0)
>     #centerline   
>     centerline_voxels = torch.tensor(skeletonize_3d(points_voxels)).long()
>     centerline_points = np.argwhere(centerline_voxels).T
>     number = 360
>     bg_mask = torch.zeros(number , points.shape[0],3)
>     depthes = torch.zeros(number , 128,128)
>     cl_mask = torch.zeros(number , centerline_points.shape[0],3)
>     projs_all = torch.zeros(number , 128, 128)
>     points_centered = points - voxel_center
>     centerline_points_centered = centerline_points - voxel_center
>     #rotate resampled_mask
>     points_image = volume2pointsplusplus(image_mha)
>     points_image_center = points_image[:,:3] - voxel_center
>     points_value = points_image[:, 3].unsqueeze(1)
>     volume = torch.zeros(number , 1, 128, 128, 128, dtype=points_image.dtype, device=points_image.device)
>     from multiprocessing import Pool
>     def process(i):
>         # Perform rotation operations on the point cloud
>         bg_mask = (rotate_coordinate_system(points_centered, i) + voxel_center) / 128
>         cl_mask = (rotate_coordinate_system(centerline_points_centered, i) + voxel_center) / 128
>         # 将Convert the rotated point cloud back to voxels
>         points_image_rotated = rotate_coordinate_system(points_image_center, i) + voxel_center
>         points_image_rotated = torch.cat((points_image_rotated / 128, points_value), dim=1)
>         rotated_image = pointsplusplus2volume(points_image_rotated, 128)
>         projs = ct_projector.forward_project(rotated_image).cpu()
>         projs_all = torch.flip(projs.squeeze(0)[0, :, :], [1]).T
>         dep = voxel_to_depth_map(rotated_image)
>         depthes = torch.flip(dep, [1]).T
>         volume = rotated_image
>         return projs_all, bg_mask, cl_mask, depthes, volume
> ```
> 
> ` You need to carefully verify the code and check whether the intermediate outputs match the expected results.

Thank you!
